### PR TITLE
[bugfix] Move setup handling to the training type plugin

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR makes sure each plugin always call datamodule setup within a distributed setting. 

Fixes #https://github.com/PyTorchLightning/pytorch-lightning/issues/8872

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1177,16 +1178,18 @@ def __set_ckpt_path(self, ckpt_path: Optional[str], model_provided: bool, model_
             )
         return ckpt_path
 
-    def _call_setup_hook(self) -> None:
+    def _call_setup_model_hook(self, model: "pl.LightningModule") -> None:
         fn = self.state.fn._setup_fn
-
         self.accelerator.barrier("pre_setup")
+        self.setup(model, stage=fn)
```
Comment (yifuwang) [inline pytorch_lightning/trainer/trainer.py:1184]: nit: maybe reflect in the method name that this also invokes the trainer `setup` hooks?

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 2]
```diff
@@ -258,6 +258,8 @@ def setup_distributed(self):
         self.dist.rank = self.global_rank
         self.dist.device = self.root_device
 
+        trainer._call_setup_datamodule_hook()
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp.py:261]: @tchaton could you describe why calling `datamodule.setup` in the trainer after `setup_environment` doesn't work? i'm concerned that having the lightning module and datamodule setups being called in two very different places will lead to other bugs down the line

**Comment by awaelchli** [comment]
@tchaton isn't the problem in #8872 that deepspeed calls `model.train_dataloader()` arbitrarily? In PL, the trainer is usually the one that accesses the dataloader through the attach and reload functions. Therefore, I cannot see how this PR can possibly fix that by moving setup(), which will affect all plugins and will be backward incompatible. Also, we lose again hook call order consistency.

**Comment by justusschock** [comment]
I agree with @awaelchli here. IMO we should go along with #6030 and if we resolve the data as early as possible, we should also be save with #8872 if I understood it correctly

**Comment by tchaton** [comment]
> @tchaton isn't the problem in #8872 that deepspeed calls `model.train_dataloader()` arbitrarily? In PL, the trainer is normally the one that accesses the dataloader through the attach and reload functions. Therefore, I cannot see how this PR can possibly fix that by moving setup(), which will affect all plugins and will be backward incompatible. Also, we lose again hook call order consistency.
> 
> If a plugin calls hooks at arbitrary places, it is only appropriate that the plugin documents this (such as the case with model.train_dataloader() reported)

Hey @awaelchli @justusschock @ananthsub,

I agree there is no perfect way to resolve this and yes DeepSpeed plugin is hacking around to access the batch_size, however I believe it is important to make sure setup is always called within a progress group initialised which is currently not the case for `ddp_spawn` and `DeepSpeed`.

Best,
T.C
